### ISSUE TITLE
Countries and memory usage

### ISSUE THREAD

**Issue by aug-riedinger** [hexorx/countries#230]
I really like the idea of not creating a specific Active Record Model for Country being static. For this reason, I find the gem really useful.

I am a little bit worried though about how much memory this gem needs. Indeed, this is often a bottleneck to an app, and it could be smarter to save RAM and to use DB's hard disk instead. 

It would be really interesting to have some insights on this issue, or even better, a profiling comparison.

Not that the gem would be useless if it were highly memory consumming, but it would be something to take into account when choosing to use it!

Thanks besides, great gem!

**Comment by rposborne** [comment]
This really boils down to speed, The countries gem becomes slow when we do some things from disk.  Mainly searching, as instead of spending the time on app boot we spend the time at call time, as to perform a meaningful search across yaml files we need to load them regardless.  

Yes disk is cheap, but speed is much more costly than memory.  Just my 2 cents.

**Comment by monfresh** [comment]
I'm also worried about the high memory usage. In one of the apps I'm working on, this gem is by far the largest memory consumer:
```
countries/country: 25.3594 MiB
```
I determined this by using the [derailed_benchmarks](https://github.com/schneems/derailed_benchmarks) gem, and in their README, they suggest opening an issue against the gems who consume an abnormal amount of memory.

**Comment by monfresh** [comment]
Here are some more details, using 0.11.5 (I ran against 0.11.3 the first time). Seems like the currencies gem plays a big part as well.

```
allocated memory by gem
-----------------------------------
    230598  countries-0.11.5
    138282  currencies-0.4.2
     18494  i18n_data-0.7.0

allocated memory by file
-----------------------------------
    180687  /Users/moncefbelyamani/.rvm/gems/ruby-2.2.2@save-ferris/gems/countries-0.11.5/lib/iso3166.rb
     99014  /Users/moncefbelyamani/.rvm/gems/ruby-2.2.2@save-ferris/gems/currencies-0.4.2/lib/iso4217.rb
     36285  /Users/moncefbelyamani/.rvm/gems/ruby-2.2.2@save-ferris/gems/currencies-0.4.2/lib/currencies/currency.rb
     31663  /Users/moncefbelyamani/.rvm/gems/ruby-2.2.2@save-ferris/gems/countries-0.11.5/lib/countries.rb
     17550  /Users/moncefbelyamani/.rvm/gems/ruby-2.2.2@save-ferris/gems/i18n_data-0.7.0/lib/i18n_data.rb
     15648  /Users/moncefbelyamani/.rvm/gems/ruby-2.2.2@save-ferris/gems/countries-0.11.5/lib/countries/country.rb
       656  /Users/moncefbelyamani/.rvm/gems/ruby-2.2.2@save-ferris/gems/currencies-0.4.2/lib/currencies/extentions.rb
      1656  /Users/moncefbelyamani/.rvm/gems/ruby-2.2.2@save-ferris/gems/countries-0.11.5/lib/countries/setup.rb
      1327  /Users/moncefbelyamani/.rvm/gems/ruby-2.2.2@save-ferris/gems/currencies-0.4.2/lib/currencies/exchange_bank.rb
       944  /Users/moncefbelyamani/.rvm/gems/ruby-2.2.2@save-ferris/gems/i18n_data-0.7.0/lib/i18n_data/version.rb
       944  /Users/moncefbelyamani/.rvm/gems/ruby-2.2.2@save-ferris/gems/countries-0.11.5/lib/countries/version.rb

allocated memory by location
-----------------------------------
     73405  /Users/moncefbelyamani/.rvm/gems/ruby-2.2.2@save-ferris/gems/countries-0.11.5/lib/iso3166.rb:6
     51255  /Users/moncefbelyamani/.rvm/gems/ruby-2.2.2@save-ferris/gems/currencies-0.4.2/lib/iso4217.rb:7
     42853  /Users/moncefbelyamani/.rvm/gems/ruby-2.2.2@save-ferris/gems/countries-0.11.5/lib/iso3166.rb:3
     29601  /Users/moncefbelyamani/.rvm/gems/ruby-2.2.2@save-ferris/gems/currencies-0.4.2/lib/iso4217.rb:8
     29454  /Users/moncefbelyamani/.rvm/gems/ruby-2.2.2@save-ferris/gems/bundler-1.10.5/lib/bundler/runtime.rb:76
     28453  /Users/moncefbelyamani/.rvm/gems/ruby-2.2.2@save-ferris/gems/countries-0.11.5/lib/iso3166.rb:5
     19641  /Users/moncefbelyamani/.rvm/gems/ruby-2.2.2@save-ferris/gems/countries-0.11.5/lib/iso3166.rb:1
     17834  /Users/moncefbelyamani/.rvm/gems/ruby-2.2.2@save-ferris/gems/currencies-0.4.2/lib/iso4217.rb:6
     16432  /Users/moncefbelyamani/.rvm/gems/ruby-2.2.2@save-ferris/gems/countries-0.11.5/lib/countries.rb:3
     16335  /Users/moncefbelyamani/.rvm/gems/ruby-2.2.2@save-ferris/gems/countries-0.11.5/lib/iso3166.rb:2
```

**Comment by monfresh** [comment]
I'm using the countries gem via [phony_rails](https://github.com/joost/phony_rails), which doesn't need the currencies portion, but because countries includes it as a dependency by default, it gets loaded in memory. Would you consider making currencies optional?

**Comment by monfresh** [comment]
Yeah, looking in more detail at the benchmark, output, it looks like the currencies gem is the real culprit here. It would really be very helpful if countries didn't include `currencies` as a default dependency, and instead, pointed users to install it separately if they need it.

**Comment by rposborne** [comment]
@monfresh awesome,  first for profiling and finding the issue.  Yes, I agree currencies should be split out, I don't have rights to tweak currencies but i think some modularization would be awesome, there are a couple parts of this gem that could be broken up or not auto loaded. 

This would need to be a major revision though as we will need to add deprecations notices on how to retain behavior.   I don't have a ton of time for OSS, over the next week or so, so if you want to hack on something feel free and I can review.

**Comment by monfresh** [comment]
I poked around the `phony_rails` gem some more, and I noticed that the only reason they use the `countries` gem is to map the 2-letter country codes to the numerical `country_code`. I just submitted a PR there to drop the `countries` dependency and just use a YAML file instead, as I think that makes more sense.

https://github.com/joost/phony_rails/pull/109

**Comment by rposborne** [comment]
We have addressed this to some extent in version 1.0.0
